### PR TITLE
Rename transport parameter initial_max_paths to initial_max_path_id

### PR DESCRIPTION
**Pull request by Yanmei-Liu**
This PR try to solve the issue #324 and remove the outdated text.

As @martinthomson pointed out in PR #372, we don't have to use a count. This would help remove the ambiguity in the old draft version: 
- When endpoints negotiate a transport parameter value with N, but initial path(Path ID = 0) is also count into the number, the max available path id is actually N-1. It's weird and doesn't match the description of Max Path ID field inside MAX_PATH frame.

### CODE REVIEW

File: draft-ietf-quic-multipath.md [hunk 1]
```diff
@@ -252,41 +252,39 @@ the use of the multipath extension during the connection handshake,
...
-The initial_max_paths parameter MUST NOT be remembered
+The initial_max_path_id parameter MUST NOT be remembered
 ({{Section 7.4.1 of QUIC-TRANSPORT}}).
 New paths can only be used after handshake completion.
 
 This extension does not change the definition of any transport parameter
 defined in {{Section 18.2. of QUIC-TRANSPORT}}.
 
-The initial_max_paths transport parameter limits the initial maximum number of active paths
-that can be used during a connection.
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:286]: Why is this removed? This is still true.

File: draft-ietf-quic-multipath.md [hunk 2]
```diff
@@ -252,41 +252,39 @@ the use of the multipath extension during the connection handshake,
...
+- initial_max_path_id (current version uses 0x0f739bbc1b666d09): the
+  initial_max_path_id transport parameter is included if the endpoint supports
   the multipath extension as defined in this document. This is
-  a variable-length integer value specifying the maximum number of
-  active concurrent paths an endpoint is willing to maintain.
-  The value of the initial_max_paths parameter MUST be at least 2.
+  a variable-length integer specifying the initial maximum value of
+  available path identifier an endpoint is willing to support.
+  The value of the initial_max_path_id parameter MUST be at least 2.
   An endpoint that receives a value less than 2 MUST close
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:261]:
```suggestion
   a variable-length integer specifying the maximum path identifier
  an endpoint is willing to maintain at connection initiation.
  E.g. if initial_max_path_id is set to 1, only connection IDs associated to Path IDs 0 and 1
  should be issued by the peer.
  The value of the initial_max_path_id parameter MUST be at least 1.
  An endpoint that receives a value less than 1 MUST close
```

File: draft-ietf-quic-multipath.md [hunk 3]
```diff
@@ -252,41 +252,39 @@ the use of the multipath extension during the connection handshake,
...
   the connection with an error of type TRANSPORT_PARAMETER_ERROR.
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:262]: However, I'm not sure if we really need to close the connection if initial_max_path_id=0. I think this would simply mean that you don't allow for any additional path yet.

Comment (Yanmei-Liu) [reply draft-ietf-quic-multipath.md:264]: If endpoints don't want to allow additional paths, endpoints could choose not to send the transport parameter. 
There's only one extra scenario: Endpoint doesn't want to allow any additional paths when handshake is finished, but keep the possibility to allow for new paths with MAX_PATH_ID frame when necessary. That could be a possible strategy.

Comment (Yanmei-Liu) [reply draft-ietf-quic-multipath.md:264]: I record this issue in #381 . If people want to add this mechanism, we could add it in a separate PR.

**Comment by Yanmei-Liu** [comment]
I'm convinced by people in issue #381 that we could allow the initial_max_path_id to be set to 0. Updated the PR. 
Thanks to @mirjak @marten-seemann and @huitema.

File: draft-ietf-quic-multipath.md [hunk 4]
```diff
@@ -1368,7 +1366,7 @@ the "QUIC Transport Parameters" registry under the "QUIC Protocol" heading.
 
 Value                                         | Parameter Name.   | Specification
 ----------------------------------------------|-------------------|-----------------
-TBD (current version uses 0x0f739bbc1b666d07) | initial_max_paths | {{nego}}
+TBD (current version uses 0x0f739bbc1b666d07) | initial_max_path_id | {{nego}}
 {: #transport-parameters title="Addition to QUIC Transport Parameters Entries"}
```
Comment (huitema) [inline draft-ietf-quic-multipath.md:1378]: I think we should change the parameter ID. The previous value meant "max path = 1 means path_id < 1"; we are changing that to "max path = 1 means path_id <= 1". That's an incompatible change, thus we need a new parameter ID value.

File: draft-ietf-quic-multipath.md [hunk 5]
```diff
@@ -1047,7 +1045,7 @@ it MUST close the connection with an error of type FRAME_ENCODING_ERROR.
 
 Receipt of multipath-specific frames
 that use a Path ID that is greater than the announced Maximum Paths value
-in the MAX_PATHS frame or in the initial_max_paths transport parameter,
+in the MAX_PATHS frame or in the initial_max_path_id transport parameter,
 if no MAX_PATHS frame was received yet,
 MUST be treated as a connection error of type MP_PROTOCOL_VIOLATION.
```
Comment (huitema) [inline draft-ietf-quic-multipath.md:1058]: To be consistent, we should we also change the name of that frame to MAX_PATH_ID and call the variable the "Maximum Path ID".

**Review by huitema — COMMENTED** [review]
If we change the transport parameter name and behavior, we need to change the parameter ID as well. We probably also need to rename the MAX_PATHS frame to MAX_PATH_ID, and change the text accordingly. And get a new frame ID for the MAX_PATH_ID frame.

File: draft-ietf-quic-multipath.md (hunk 4 above)
Comment (Yanmei-Liu) [reply draft-ietf-quic-multipath.md:1378]: It's true. I've changed the parameter ID from 0x0f739bbc1b666d07 to 0x0f739bbc1b666d09.

File: draft-ietf-quic-multipath.md (hunk 5 above)
Comment (Yanmei-Liu) [reply draft-ietf-quic-multipath.md:1058]: That's true. I updated the Frame Name and Type from MAX_PATHS to MAX_PATH_ID, please review.

Comment (Yanmei-Liu) [reply draft-ietf-quic-multipath.md:1058]: Note that we only need to change Frame Type and Frame Name in this PR, because Martin has already changed the definition of Frame field "Maximum Path ID" in PR #372.

File: draft-ietf-quic-multipath.md [hunk 6]
```diff
@@ -1308,38 +1314,38 @@ As the MP_NEW_CONNECTION_ID frames applies the sequence number per path,
...
 
-MAX_PATHS frames are formatted as shown in {{fig-max-paths-frame-format}}.
+MAX_PATH_ID frames are formatted as shown in {{fig-max-paths-frame-format}}.
 
 ~~~
-MAX_PATHS Frame {
-  Type (i) = 0x15228c0b,
+MAX_PATH_ID Frame {
+  Type (i) = 0x15228c0c,
   Maximum Paths (i),
```
Comment (martinthomson) [inline draft-ietf-quic-multipath.md:1330]: This probably needs to be "Maximum Path ID" (and the same throughout)

File: draft-ietf-quic-multipath.md [hunk 7]
```diff
@@ -1308,42 +1314,43 @@ As the MP_NEW_CONNECTION_ID frames applies the sequence number per path,
...
 ~~~
-{: #fig-max-paths-frame-format title="MAX_PATHS Frame Format"}
+{: #fig-max-paths-frame-format title="MAX_PATH_ID Frame Format"}
 
-MAX_PATHS frames contain the following field:
+MAX_PATH_ID frames contain the following field:
 
 Maximum Path Identifier:
 : A count of the cumulative number of paths that can be opened
   over the lifetime of the connection. This value MUST NOT exceed 2^32-1, as
```
Comment (martinthomson) [inline draft-ietf-quic-multipath.md:1339]: This definition might be better stated as:

> The maximum path identifier that the sending endpoint is willing to accept.

File: draft-ietf-quic-multipath.md [hunk 8]
```diff
@@ -1308,42 +1314,43 @@ As the MP_NEW_CONNECTION_ID frames applies the sequence number per path,
...
   Path IDs are defined with a maximum value 2^32-1 as the 32 bits of the Path ID are used
-  to calculate the nonce (see Section {{multipath-aead}}).
+  to calculate the nonce (see {{multipath-aead}}).
   The Maximum Paths value MUST NOT be lower than the value
```
Comment (martinthomson) [inline draft-ietf-quic-multipath.md:1342]:
```suggestion
  The Maximum Path Identifier value MUST NOT be lower than the value
```

File: draft-ietf-quic-multipath.md [hunk 9]
```diff
@@ -1308,42 +1314,43 @@ As the MP_NEW_CONNECTION_ID frames applies the sequence number per path,
...
-  advertised in the initial_max_paths transport parameter. Receipt
-  of an invalid Maximum Paths value MUST be treated as a
+  advertised in the initial_max_path_id transport parameter. Receipt
```
Comment (martinthomson) [inline draft-ietf-quic-multipath.md:1343]: ... or what?  Is it a frame error to receive a MAX_PATH_ID with a value less than (or equal to...?) the transport parameter?

**Review by martinthomson — APPROVED** [review]
This seems good.

File: draft-ietf-quic-multipath.md (hunk 7 above)
Comment (Yanmei-Liu) [reply draft-ietf-quic-multipath.md]: That's true. I've updated the definition of maximum path identifier field.

File: draft-ietf-quic-multipath.md [hunk 10]
```diff
@@ -252,41 +252,47 @@ the use of the multipath extension during the connection handshake,
...
-Setting initial_max_paths parameter is equivalent to sending a
-MAX_PATHS frame ({{max-paths-frame}}) with the same value.
-
-If either of the endpoints does not advertise the initial_max_paths transport
+  a variable-length integer specifying the maximum path identifier
+  an endpoint is willing to maintain at connection initiation.
+  For example, if initial_max_path_id is set to 1, only connection IDs
+  associated with Path IDs 0 and 1 should be issued by the peer.
+  If an endpoint receives an initial_max_path_id transport parameter with value 0,
+  it MUST NOT cause any error, and notice that the peer want to
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:263]:
```suggestion
the peer want to
```

File: draft-ietf-quic-multipath.md [hunk 11]
```diff
@@ -252,41 +252,47 @@ the use of the multipath extension during the connection handshake,
...
+  enable the multipath extension without allowing extra paths immediately
+  after the handshake is done. If initial_max_path_id is successfully negotiated
+  (including negotiated as 0), endpoints can use MAX_PATH_ID frame to increase
+  the maximum path identifier and enable the use of new paths.
```
Comment (mirjak) [inline draft-ietf-quic-multipath.md:267]:
```suggestion
the peer aims to  enable the multipath extension without allowing extra paths immediately.
To allow for the use of more paths later,
endpoints can send the MAX_PATH_ID frame to increase the maximum allowed path identifier.
```

**Review by huitema — APPROVED** [review]
I think this is ready.

File: draft-ietf-quic-multipath.md [hunk 12]
```diff
@@ -252,41 +252,45 @@ the use of the multipath extension during the connection handshake,
...
-
-If either of the endpoints does not advertise the initial_max_paths transport
+  a variable-length integer specifying the maximum path identifier
+  an endpoint is willing to maintain at connection initiation.
+  For example, if initial_max_path_id is set to 1, only connection IDs
+  associated with Path IDs 0 and 1 should be issued by the peer.
+  If an endpoint receives an initial_max_path_id transport parameter with value 0,
+the peer aims to  enable the multipath extension without allowing extra paths immediately.
+To allow for the use of more paths later,
+endpoints can send the MAX_PATH_ID frame to increase the maximum allowed path identifier.
```
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:265]:
```suggestion
  the peer aims to  enable the multipath extension without allowing extra paths immediately.
  To allow for the use of more paths later,
  endpoints can send the MAX_PATH_ID frame to increase the maximum allowed path identifier.
```

**Review by qdeconinck — APPROVED** [review]
Modulo a formatting issue, this PR is ready to merge.
